From 943398ff9a266de0c1c1939d3454a47b4195f68f Mon Sep 17 00:00:00 2001
From: Meriipu <Meriipu@users.noreply.github.com>
Date: Mon, 12 Oct 2020 02:18:42 +0200
Subject: [PATCH 4/9] revert the removal of the dropmarker et al.

as of 78 this alone is not enough to have the dropmarker be functional
some additional patches to fix autoOpen on focus/select as well as
avoiding one of the providers is necessary. This commit alone only adds
the dropmarker back cosmetically.

Hardcode tooltiptext instead of dealing with i10l (mozilla removed the
entries and keep changing up how they are organized so so they have to
be added back manually, which is unnecessary effort since the
translations likely are removed as well).  Mozilla changing this
organization up appears to be the cause of the occasional error:
  `XML Parsing Error: undefined entity`
when starting Firefox after applying the patches to recent releases.
---
 browser/base/content/browser.css              |  1 +
 browser/base/content/browser.xhtml            |  3 +++
 browser/components/urlbar/UrlbarInput.jsm     | 19 +++++++++++++++
 browser/components/urlbar/UrlbarView.jsm      |  2 ++
 .../themes/shared/urlbar-searchbar.inc.css    | 24 +++++++++++++++++++
 5 files changed, 49 insertions(+)

diff --git a/browser/base/content/browser.css b/browser/base/content/browser.css
index 085e47cad9f0..45fc6d7f10bd 100644
--- a/browser/base/content/browser.css
+++ b/browser/base/content/browser.css
@@ -704,6 +704,7 @@ toolbar:not(#TabsToolbar) > #personal-bookmarks {
 #urlbar-input-container[pageproxystate=invalid] > #page-action-buttons > .urlbar-page-action,
 #identity-box.chromeUI ~ #page-action-buttons > .urlbar-page-action:not(#star-button-box),
 #urlbar[usertyping] > #urlbar-input-container > #page-action-buttons > #urlbar-zoom-button,
+#urlbar[usertyping] > #urlbar-input-container > .urlbar-history-dropmarker,
 #urlbar:not([usertyping]) > #urlbar-input-container > #urlbar-go-button,
 #urlbar:not([focused]) > #urlbar-input-container > #urlbar-go-button {
   display: none;
diff --git a/browser/base/content/browser.xhtml b/browser/base/content/browser.xhtml
index 965fb8b98b81..ce0449986bb3 100644
--- a/browser/base/content/browser.xhtml
+++ b/browser/base/content/browser.xhtml
@@ -1877,6 +1877,9 @@
                      class="urlbar-icon"
                      onclick="gURLBar.handleCommand(event);"
                      data-l10n-id="urlbar-go-button"/>
+              <image class="urlbar-history-dropmarker urlbar-icon chromeclass-toolbar-additional"
+                     hidden="false"
+                     tooltiptext="Show history"/>
               <hbox id="page-action-buttons" context="pageActionContextMenu">
                 <toolbartabstop/>
                 <hbox id="contextual-feature-recommendation" role="button" hidden="true">
diff --git a/browser/components/urlbar/UrlbarInput.jsm b/browser/components/urlbar/UrlbarInput.jsm
index 989fb8d3b649..87416797cb23 100644
--- a/browser/components/urlbar/UrlbarInput.jsm
+++ b/browser/components/urlbar/UrlbarInput.jsm
@@ -155,6 +155,8 @@ class UrlbarInput {
     }
 
     this.inputField = this.querySelector("#urlbar-input");
+    this.dropmarker = this.querySelector(".urlbar-history-dropmarker");
+    this.dropmarker.hidden = false;
     this._inputContainer = this.querySelector("#urlbar-input-container");
     this._identityBox = this.querySelector("#identity-box");
     this._searchModeIndicator = this.querySelector(
@@ -209,6 +211,7 @@ class UrlbarInput {
       this.addEventListener(name, this);
     }
 
+    this.dropmarker.addEventListener("mousedown", this);
     this.window.addEventListener("mousedown", this);
     if (AppConstants.platform == "win") {
       this.window.addEventListener("draggableregionleftmousedown", this);
@@ -2398,6 +2401,22 @@ class UrlbarInput {
           this.view.autoOpen({ event });
         }
         break;
+      case this.dropmarker:
+        if (event.button != 0) {
+          break;
+        }
+
+        if (this.view.isOpen) {
+          this.view.close();
+        } else {
+          this.focus();
+          this.startQuery({
+            allowAutofill: false,
+            event,
+          });
+          this._maybeSelectAll();
+        }
+        break;
       case this.window:
         if (this._mousedownOnUrlbarDescendant) {
           this._mousedownOnUrlbarDescendant = false;
diff --git a/browser/components/urlbar/UrlbarView.jsm b/browser/components/urlbar/UrlbarView.jsm
index 17b0faed0808..def090cba822 100644
--- a/browser/components/urlbar/UrlbarView.jsm
+++ b/browser/components/urlbar/UrlbarView.jsm
@@ -388,6 +388,7 @@ class UrlbarView {
 
     this.removeAccessibleFocus();
     this.input.inputField.setAttribute("aria-expanded", "false");
+    this.input.dropmarker.removeAttribute("open");
 
     this.input.removeAttribute("open");
     this.input.endLayoutExtend();
@@ -743,6 +744,7 @@ class UrlbarView {
     this._enableOrDisableRowWrap();
 
     this.input.inputField.setAttribute("aria-expanded", "true");
+    this.input.dropmarker.setAttribute("open", "true");
 
     this.input.setAttribute("open", "true");
     this.input.startLayoutExtend();
diff --git a/browser/themes/shared/urlbar-searchbar.inc.css b/browser/themes/shared/urlbar-searchbar.inc.css
index 954cf74a14ad..9a2394d95857 100644
--- a/browser/themes/shared/urlbar-searchbar.inc.css
+++ b/browser/themes/shared/urlbar-searchbar.inc.css
@@ -530,6 +530,19 @@
   transform: scaleX(-1);
 }
 
+.urlbar-history-dropmarker {
+  list-style-image: url(chrome://global/skin/icons/arrow-dropdown-16.svg);
+  transition: opacity 0.15s ease;
+}
+
+#urlbar[switchingtabs] > #urlbar-input-container > .urlbar-history-dropmarker {
+  transition: none;
+}
+
+#nav-bar:not([customizing="true"]) > #nav-bar-customization-target > #urlbar-container:not(:hover) > #urlbar:not([focused]) > #urlbar-input-container > .urlbar-history-dropmarker {
+  opacity: 0;
+}
+
 #pageActionButton,
 .share-more-button {
   list-style-image: url("chrome://global/skin/icons/more.svg");
@@ -727,6 +740,17 @@
   mask-position-x: 0;
 }
 
+/* Translate the dropmarker to give illusion of increasing width of the recommendation  */
+#urlbar[cfr-recommendation-state] .urlbar-history-dropmarker {
+  transition: transform ease-in-out var(--cfr-animation-duration);
+}
+#urlbar[cfr-recommendation-state="expanded"] .urlbar-history-dropmarker {
+  transform: translateX(calc(var(--cfr-label-width) * -1));
+}
+#urlbar[cfr-recommendation-state="expanded"] .urlbar-history-dropmarker:-moz-locale-dir(rtl) {
+  transform: translateX(calc(var(--cfr-label-width)));
+}
+
 /* Shift the url-bar text fading to stop the recommendation overlapping */
 #urlbar[cfr-recommendation-state] #urlbar-input {
   /* A mask-image is usually only defined for the url bar when text overflows.
-- 
2.28.0

